blockchain/chain.go

```diff
@@ -1679,6 +1674,156 @@ func (b *BlockChain) HeaderByHash(hash *chainhash.Hash) (wire.BlockHeader, error
+// the main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) MainChainHasBlock(hash *chainhash.Hash) bool {
+	node := b.index.LookupNode(hash)
+	b.chainLock.RLock()
+	hasBlock := node != nil && node.inMainChain
+	b.chainLock.RUnlock()
+	return hasBlock
+}
+
+// BlockHeightByHash returns the height of the block with the given hash in the
+// main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockHeightByHash(hash *chainhash.Hash) (int64, error) {
+	node := b.index.LookupNode(hash)
+	b.chainLock.RLock()
+	if node == nil || !node.inMainChain {
+		b.chainLock.RUnlock()
+		str := fmt.Sprintf("block %s is not in the main chain", hash)
+		return 0, errNotInMainChain(str)
+	}
+	b.chainLock.RUnlock()
+
+	return node.height, nil
+}
+
+// BlockHashByHeight returns the hash of the block at the given height in the
+// main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockHashByHeight(height int64) (*chainhash.Hash, error) {
+	b.heightLock.RLock()
+	node := b.mainNodesByHeight[height]
+	b.heightLock.RUnlock()
+	if node == nil {
+		str := fmt.Sprintf("no block at height %d exists", height)
+		return nil, errNotInMainChain(str)
+	}
+
+	return &node.hash, nil
+}
+
+// HeightRange returns a range of block hashes for the given start and end
+// heights.  It is inclusive of the start height and exclusive of the end
+// height.  The end height will be limited to the current main chain height.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) HeightRange(startHeight, endHeight int64) ([]chainhash.Hash, error) {
+	// Ensure requested heights are sane.
+	if startHeight < 0 {
+		return nil, fmt.Errorf("start height of fetch range must not "+
+			"be less than zero - got %d", startHeight)
+	}
+	if endHeight < startHeight {
+		return nil, fmt.Errorf("end height of fetch range must not "+
+			"be less than the start height - got start %d, end %d",
+			startHeight, endHeight)
+	}
+
+	// There is nothing to do when the start and end heights are the same,
+	// so return now to avoid the chain lock.
+	if startHeight == endHeight {
+		return nil, nil
+	}
+
+	// Grab a lock on the chain to prevent it from changing due to a reorg
+	// while building the hashes.
+	b.chainLock.RLock()
+	defer b.chainLock.RUnlock()
+
+	// When the requested start height is after the most recent best chain
+	// height, there is nothing to do.
+	tip := b.bestNode
+	latestHeight := tip.height
+	if startHeight > latestHeight {
+		return nil, nil
+	}
+
+	// Limit the ending height to the latest height of the chain.
+	if endHeight > latestHeight+1 {
+		endHeight = latestHeight + 1
+	}
+
+	// Fetch requested hashes.
+	hashes := make([]chainhash.Hash, endHeight-startHeight)
+	b.heightLock.RLock()
+	iterNode := b.mainNodesByHeight[endHeight-1]
+	b.heightLock.RUnlock()
+	for i := startHeight; i < endHeight; i++ {
```

Comment:
this would read better to me if it iterated backwards until reaching the start height

Reply:
I'd rather not change this since it has already been well tested manually and reversing it could easily introduce an off by one.  Since the blocklocator bits no longer use this function, the only remaining caller is from RPC handlers for which there are, unfortunately, no automated tests.